packages/node/src/indexer/dictionary/v2/ethDictionaryV2.ts

```diff
@@ -0,0 +1,293 @@
+            filter.function
+          ) {
+            fatDictionaryConditions.transactions.push(
+              callFilterToFatDictionaryCondition(filter, ds.options),
+            );
+          } else {
+            // do nothing;
+          }
+          break;
+        }
+        case EthereumHandlerKind.Event: {
+          const filter = handler.filter as EthereumLogFilter;
+          if (ds.groupedOptions) {
+            fatDictionaryConditions.logs.push(
+              eventFilterToFatDictionaryCondition(filter, ds.groupedOptions),
+            );
+          } else if (ds.options?.address || filter.topics) {
+            fatDictionaryConditions.logs.push(
+              eventFilterToFatDictionaryCondition(filter, ds.options),
+            );
+          } else {
+            // do nothing;
+          }
+          break;
+        }
+        default:
+      }
+    }
+  }
+
+  //TODO, unique
+  return fatDictionaryConditions;
+  // return uniqBy(
+  //   allDictionaryConditions,
+  //   (item) =>
+  //     `${item}|${JSON.stringify(
+  //       sortBy(item.conditions, (c) => c.field),
+  //     )}`,
+  // );
+}
+
+export class EthDictionaryV2 extends DictionaryV2<
+  EthereumBlock,
+  SubqlDatasource,
+  EthDictionaryV2QueryEntry
+> {
+  protected buildDictionaryQueryEntries(
+    dataSources: SubqlDatasource[],
+  ): DictionaryV2QueryEntry {
+    const filteredDs = ethFilterDs(dataSources);
+    return buildDictionaryV2QueryEntry(filteredDs);
+  }
+
+  constructor(
+    endpoint: string,
+    nodeConfig: NodeConfig,
+    eventEmitter: EventEmitter2,
+    project: SubqueryProject,
+    chainId?: string,
+  ) {
+    super(
+      endpoint,
+      chainId ?? project.network.chainId,
+      nodeConfig,
+      eventEmitter,
+    );
+  }
+
+  /**
+   *
+   * @param startBlock
+   * @param queryEndBlock this block number will limit the max query range, increase dictionary query speed
+   * @param batchSize
+   * @param conditions
+   */
+  async getData(
+    startBlock: number,
+    queryEndBlock: number,
+    limit = MIN_FAT_FETCH_LIMIT,
+  ): Promise<DictionaryResponse<IBlock<EthereumBlock>> | undefined> {
+    const queryDetails = this.queriesMap?.getDetails(startBlock);
+    const conditions = queryDetails?.value;
+    queryEndBlock = this.metadata.end;
+
+    if (!conditions) {
+      return undefined;
+    }
+    const requestData = {
+      jsonrpc: '2.0',
+      method: FAT_BLOCKS_QUERY_METHOD,
+      id: 1,
+      params: [
+        startBlock,
+        queryEndBlock,
+        limit,
+        conditions,
+        { blockHeader: true, logs: true, transactions: { data: true } },
+      ],
+    };
+
+    try {
+      const response = await this.dictionaryApi.post(
+        this.dictionaryEndpoint,
+        requestData,
+        {
+          headers: {
+            'Content-Type': 'application/json',
+          },
+        },
+      );
+      const ethBlocks = this.convertResponseBlocks(response.data.result);
```

Comment:
This function can return undefined, if that happens then the return will fail. If the result is empty then `lastBufferedHeight` should be `queryEndBlock`